src/mds/Server.cc

```diff
@@ -4153,31 +4153,15 @@ void Server::handle_client_getattr(const MDRequestRef& mdr, bool is_lookup)
     } else if (ref->filelock.is_stable() ||
 	       ref->filelock.get_num_wrlocks() > 0 ||
 	       !ref->filelock.can_read(mdr->get_client())) {
-      /* Since we're taking advantage of an optimization here:
-       *
-       * We cannot suddenly, due to a changing condition, add this filelock as
-       * it can cause lock-order deadlocks. In this case, that condition is the
-       * lock state changes between request retries. If that happens, we need
-       * to check if we've acquired the other locks in this vector. If we have,
-       * then we need to drop those locks and retry.
-       */
-      if (mdr->is_rdlocked(&ref->linklock) ||
-          mdr->is_rdlocked(&ref->authlock) ||
-          mdr->is_rdlocked(&ref->xattrlock)) {
-        /* start over */
-        dout(20) << " dropping locks and restarting request because filelock state change" << dendl;
-	mds->locker->drop_locks(mdr.get());
-	mdr->drop_local_auth_pins();
-	mds->queue_waiter(new C_MDS_RetryRequest(mdcache, mdr));
-        return;
-      }
       lov.add_rdlock(&ref->filelock);
       mdr->locking_state &= ~MutationImpl::ALL_LOCKED;
     }
   }
 
-  if (!mds->locker->acquire_locks(mdr, lov))
+  if (!mds->locker->acquire_locks(mdr, lov)) {
+    restart_over_the_request(mdr);
```

Comment:
This doesn't look right. Anytime the Locker is nudging a lock state (perhaps due to conflicting caps) then we will drop all locks. That will prevent the request from making progress.

Reply:
Then I need to do this in another way and maybe more complex.

Reply:
What you describe is a deadlock, though not because of a lack of interfaces, but rather due to the bad locking order by req1 and/or req2.

We should be careful updating lock orders in bulk, since the current stable state may be kicked out of balance, even if there are a few places where it would deadlock today.

in general, requests should try and take all the locks they need in a single call to `acquire_locks`, unless they have multiple phases. In the latter case, the request should be responsible of releasing locks from phase 1 before re-acquiring them in phase 2 along with other locks. The only case when request may keep locks from phase 1 to phase 2 is if they are ordered before the new locks from phase 2.

`req1` violates this by taking INEST while holding IFILE. Whether `req1` can afford dropping and then re-acquiring INEST along with IFILE and IXATTR is not known, but it's `req1`'s responsibility to never try to lock backwards in that defined lock order

Reply:
> What you describe is a deadlock, though not because of a lack of interfaces, but rather due to the bad locking order by req1 and/or req2.
> 
> We should be careful updating lock orders in bulk, since the current stable state may be kicked out of balance, even if there are a few places where it would deadlock today.
> 
> in general, requests should try and take all the locks they need in a single call to `acquire_locks`, unless they have multiple phases. In the latter case, the request should be responsible for releasing locks from phase 1 before re-acquiring them in phase 2 along with other locks. The only case when a request may keep locks from phase 1 to phase 2 is if they are ordered before the new locks from phase 2.
> 
> `req1` violates this by taking INEST while holding IFILE. Whether `req1` can afford dropping and then re-acquiring INEST along with IFILE and IXATTR is not known, but it's `req1`'s responsibility to never try to lock backwards in that defined lock order. Note that it is OK for `req1` to continue holding ISNAP and IFILE, since no new locks are ordered before those two

Unfortunately there are many cases will call the `acquire_locks()` twice or more. Such as for the `getattr` request it will do it first when traversing the path in `traverse_path()` and then in the `_getattr()` func. And this really will mess the locks order.

It's not that easy to merge them in a single `acquire_locks()` call.

Reply:
Yes, understood, which is why I mentioned those multi-phase requests, where each phase takes additional locks. Those should be designed to only acquire locks in ascending order